addons/binding/org.openhab.binding.plclogo/src/main/java/org/openhab/binding/plclogo/handler/PLCDigitalHandler.java

```diff
@@ -0,0 +1,325 @@
+            }
+        } else {
+            logger.warn("Invalid channel {} found.", channelUID);
+        }
+    }
+
+    @Override
+    public void setData(final byte[] data) {
+        if (!isThingOnline()) {
+            return;
+        }
+
+        if (data.length != getBufferLength()) {
+            logger.warn("Received and configured data sizes does not match.");
+            return;
+        }
+
+        final List<Channel> channels = thing.getChannels();
+        if (channels.size() != getNumberOfChannels()) {
+            logger.warn("Received and configured channels sizes does not match.");
+            return;
+        }
+
+        for (final Channel channel : channels) {
+            final ChannelUID channelUID = channel.getUID();
+            final String name = getBlockFromChannel(channel);
+
+            final int bit = getBit(name);
+            final int address = getAddress(name);
+            if ((address != INVALID) && (bit != INVALID)) {
+                final DecimalType state = (DecimalType) getOldValue(name);
+                boolean value = S7.GetBitAt(data, address - getBase(name), bit);
+                if ((state == null) || ((value ? 1 : 0) != state.intValue()) || config.isUpdateForced()) {
+                    updateChannel(channel, value);
+                }
+            } else {
+                logger.warn("Invalid channel {} found.", channelUID);
+            }
+        }
+
+        /*
+         * if (logger.isTraceEnabled()) {
+         * final String raw = "[" + Integer.toBinaryString((data[0] & 0xFF) + 0x100).substring(1) + "]";
+         * logger.trace("Channel {} accepting {} received {}.", channel.getUID(), type, raw);
+         * }
+         */
+    }
+
+    @Override
+    protected void updateState(ChannelUID channelUID, State state) {
+        super.updateState(channelUID, state);
+        DecimalType value = state.as(DecimalType.class);
+        if (state instanceof OpenClosedType) {
+            final OpenClosedType type = (OpenClosedType) state;
+            value = new DecimalType(type == OpenClosedType.CLOSED ? 1 : 0);
+        }
+
+        final Channel channel = thing.getChannel(channelUID.getId());
+        Objects.requireNonNull(channel, "PLCDigitalHandler: Invalid channel found.");
+
+        setOldValue(getBlockFromChannel(channel), value);
+    }
+
+    @Override
+    protected void updateConfiguration(Configuration configuration) {
+        super.updateConfiguration(configuration);
+        synchronized (config) {
+            config = getConfigAs(PLCDigitalConfiguration.class);
+        }
+    }
+
+    @Override
+    protected boolean isValid(final String name) {
+        if (2 <= name.length() && (name.length() <= 4)) {
+            final String kind = getBlockKind();
+            if (Character.isDigit(name.charAt(1)) || Character.isDigit(name.charAt(2))) {
+                boolean valid = kind.equalsIgnoreCase("I") || kind.equalsIgnoreCase("NI");
```

Comment:
Position literals first in `String` comparisons - that way if the `String` is `null` you won't get a `NullPointerException`, it'll just return `false`. [source: PMD](http://pmd.sourceforge.net/pmd-4.3.0/rules/design.html#PositionLiteralsFirstInComparisons)

Please check all